diff --git a/setup/setup_database.py b/setup/setup_database.py
index ec88437a..10751c8a 100755
--- a/setup/setup_database.py
+++ b/setup/setup_database.py
@@ -40,7 +40,7 @@ elif STAGING_KEY == "RANDOM":
 
 # Calculate the install path. We know the project directory will always be the parent of the current directory. Any modifications of the folder structure will
 # need to be applied here.
-INSTALL_PATH = os.path.dirname(os.path.dirname(os.path.realpath(__file__))) + "/"
+INSTALL_PATH = os.get_env('INSTALL_PATH', os.path.dirname(os.path.dirname(os.path.realpath(__file__))) + "/")
 
 # an IP white list to ONLY accept clients from
 #   format is "192.168.1.1,192.168.1.10-192.168.1.100,10.0.0.0/8"
